File: src/subtle/gravity.c

 /**
  * @package subtle
  *
  * @file Gravity functions
  * @copyright (c) 2005-2009 Christoph Kappel <unexist@subforge.org>
  * @version $Id: src/subtle/gravity.c,v 3213 2012/05/31 13:42:04 unexist $
  *
  * This program can be distributed under the terms of the GNU GPLv2.
  * See the file COPYING for details.
  **/

#include <X11/Xresource.h>
#include "subtle.h"

/* Method */

 /** subGravityNew {{{
  * @brief Create new gravity
  * @param[in]  name  Gravity name
  * @param[in]  geom  Geometry of gravity
  * @return Returns a new #SubGravity or \p NULL
  **/

SubGravity *
subGravityNew(const char *name,
  XRectangle *geom)
{
  SubGravity *g = NULL;

  assert(name && geom);

  /* Create gravity */
  g = GRAVITY(subSharedMemoryAlloc(1, sizeof(SubGravity)));
  g->flags |= SUB_TYPE_GRAVITY;

  if(name) g->quark = XrmStringToQuark(name); ///< Create hash

  /* Sanitize values */
  g->geom.x      = MINMAX(geom->x,      0, 100);
  g->geom.y      = MINMAX(geom->y,      0, 100);
  g->geom.width  = MINMAX(geom->width,  1, 100);
  g->geom.height = MINMAX(geom->height, 1, 100);

  subSubtleLogDebugSubtle("New: name=%s, quark=%d, x=%d, y=%d,"
    "width=%d, height=%d\n",
    name, g->quark, geom->x, geom->y, geom->width, geom->height);

  return g;
} /* }}} */

 /** subGravityGeometry {{{
  * @brief Calculate geometry of gravity for bounds
  * @param[in]   g       A #SubGravity
  * @param[in]   bounds  A #XRectangle
  * @param[out]  geom    A #XRectangle
  **/

void
subGravityGeometry(SubGravity *g,
  XRectangle *bounds,
  XRectangle *geom)
{
  assert(g && bounds && geom);

  /* Calculate gravity size for bounds */
  geom->x      = bounds->x + (bounds->width * g->geom.x / 100);
  geom->y      = bounds->y + (bounds->height * g->geom.y / 100);
  geom->width  = (bounds->width * g->geom.width / 100);
  geom->height = (bounds->height * g->geom.height / 100);
} /* }}} */

 /** subGravityKill {{{
  * @brief Kill gravity
  * @param[in]  g  A #SubGravity
  **/

void
subGravityKill(SubGravity *g)
{
  assert(g);

  free(g);

  subSubtleLogDebugSubtle("Kill\n");
} /* }}} */

/* All */

 /** subGravityFind {{{
  * @brief Find gravity
  * @param[in]   name   Gravity name
  * @param[in]   quark  Quark
  * @retval  -1  Not found
  * @retval  >0  Found id
  **/

int
subGravityFind(const char *name,
  int quark)
{
  int found = -1;

  if(0 < subtle->gravities->ndata)
    {
      int i, hash = 0;

      /* Get quark */
      if(name) hash = XrmStringToQuark(name);
      else hash = quark;

      for(i = 0; i < subtle->gravities->ndata; i++)
        {
          SubGravity *g = GRAVITY(subtle->gravities->data[i]);

          /* Compare quarks */
          if(g->quark == hash)
            {
              found = i;
              break;
            }
        }
    }

  return found;
} /* }}} */

 /** subGravityPublish {{{
  * @brief Publish gravities
  **/

void
subGravityPublish(void)
{
  int i;
  char **gravities = NULL, buf[30] = { 0 };
  SubGravity *g = NULL;

  assert(0 < subtle->gravities->ndata);

  /* Alloc space */
  gravities  = (char **)subSharedMemoryAlloc(subtle->gravities->ndata,
    sizeof(char *));

  for(i = 0; i < subtle->gravities->ndata; i++)
    {
      g = GRAVITY(subtle->gravities->data[i]);

      /* Add gravity to list */
      snprintf(buf, sizeof(buf), "%dx%d+%d+%d#%s", g->geom.x,
        g->geom.y, g->geom.width, g->geom.height,
        XrmQuarkToString(g->quark));

      gravities[i] = (char *)subSharedMemoryAlloc(strlen(buf) + 1,
        sizeof(char));
      strncpy(gravities[i], buf, strlen(buf));
    }

  /* EWMH: Gravity list and geometries */
  subSharedPropertySetStrings(subtle->dpy, ROOT,
    subEwmhGet(SUB_EWMH_SUBTLE_GRAVITY_LIST), gravities,
    subtle->gravities->ndata);

  /* Tidy up */
  for(i = 0; i < subtle->gravities->ndata; i++)
    free(gravities[i]);

  XSync(subtle->dpy, False); ///< Sync all changes

  free(gravities);

  subSubtleLogDebugSubtle("Publish: gravities=%d\n", subtle->gravities->ndata);
} /* }}} */

// vim:ts=2:bs=2:sw=2:et:fdm=marker

File: test/contexts/gravity.rb
#
# @package test
#
# @file Test Subtlext::Gravity functions
# @author Christoph Kappel <unexist@subforge.org>
# @version $Id: test/contexts/gravity.rb,v 3169 2012/01/03 20:43:30 unexist $
#
# This program can be distributed under the terms of the GNU GPLv2.
# See the file COPYING for details.
#

context 'Gravity' do
  GRAVITY_COUNT = 30
  GRAVITY_ID    = 12
  GRAVITY_NAME  = 'center'

  setup do # {{{
    Subtlext::Gravity.first(GRAVITY_ID)
  end # }}}

  asserts 'Check attributes' do # {{{
    GRAVITY_ID == topic.id and GRAVITY_NAME == topic.name and
      topic.geometry.is_a? Subtlext::Geometry
  end # }}}

  asserts 'Get list' do # {{{
    list = Subtlext::Gravity.list

    list.is_a?(Array) and GRAVITY_COUNT == list.size and
      Subtlext::Gravity.method(:all) == Subtlext::Gravity.method(:list)
  end # }}}

  asserts 'Finder' do # {{{
    index  = Subtlext::Gravity[GRAVITY_ID]
    string = Subtlext::Gravity[GRAVITY_NAME + '$']
    sym    = Subtlext::Gravity[GRAVITY_NAME.to_sym]
    all    = Subtlext::Gravity.find('.*')
    none   = Subtlext::Gravity['abcdef']

    index == string and index == sym and
      all.is_a?(Array) and GRAVITY_COUNT == all.size and
      none.nil?
  end # }}}

  asserts 'First' do # {{{
    index  = Subtlext::Gravity.first(GRAVITY_ID)
    string = Subtlext::Gravity.first(GRAVITY_NAME)

    index == string
  end # }}}

  asserts 'Equal and compare' do # {{{
    topic.eql?(Subtlext::Gravity.first(GRAVITY_ID)) and topic == topic
  end # }}}

  asserts 'Hash and unique' do # {{{
    1 == [ topic, topic ].uniq.size
  end # }}}

  asserts 'Check associations' do # {{{
    clients = topic.clients

    clients.is_a?(Array) and 1 == clients.size
  end # }}}

  asserts 'Check geometry for screen' do # {{{
    screen = Subtlext::Screen.current
    geom   = topic.geometry_for(screen)

    screen.geometry == geom
  end # }}}

  asserts 'Convert to string' do # {{{
    GRAVITY_NAME == topic.to_str
  end # }}}

  asserts 'Create new gravity' do # {{{
    g = Subtlext::Gravity.new('test', 0, 0, 100, 100)
    g.save

    sleep 0.5

    GRAVITY_COUNT + 1 == Subtlext::Gravity.all.size
  end # }}}

  asserts 'Test tiling' do # {{{
    topic.tiling = :vert
    topic.tiling = :horz
    topic.tiling = nil

    true
  end # }}}

  asserts 'Kill a gravity' do # {{{
    Subtlext::Gravity.first('test').kill

    sleep 0.5

    GRAVITY_COUNT == Subtlext::Gravity.all.size
  end # }}}
end

# vim:ts=2:bs=2:sw=2:et:fdm=marker
